### ISSUE TITLE
Logit much slower than sklearn LogisticRegression

### ISSUE THREAD

**Issue by aizvorski** [statsmodels/statsmodels#6808]
#### Describe the bug

Performance bug: statsmodels Logit regression is 10-100x slower than scikit-learn LogisticRegression.

I benchmarked both using L-BFGS solver, with the same number of iterations, and the same other settings as far as I can tell.

The speed difference seems to increase with larger data sets, and increase by a huge amount when using fit_regularized().  On dataset 10k samples x 1k features, fit_regularized() with l1 penalty can take more than a day, where scikit-learn with l2 penalty takes a couple of seconds.  

On my production data, statsmodels is so slow as to be essentially unusable.

I'm also wondering if the two version of logit somehow work in a completely different way: is this only a performance bug, or a performance difference that is due to a significant difference in the way the optimization problem is posed? (and consequently with different results)

#### Code Sample, a copy-pastable example if possible

```python
import statsmodels.api as sm
import sklearn.datasets
import sklearn.linear_model # sklearn 0.22.1
X_train, y_train = sklearn.datasets.make_classification(n_samples=30000, n_features=2048, n_informative=2048, n_redundant=0, n_repeated=0)
%time model = sm.Logit(y_train, X_train).fit(method='lbfgs', pgtol=0.0001, maxiter=10, disp=True, qc_verbose=True)
# CPU times: user 37.5 s, sys: 1.4 s, total: 38.8 s Wall time: 9.98 s
%time model = sklearn.linear_model.LogisticRegression(max_iter=10, penalty='none', verbose=1).fit(X_train, y_train)
# CPU times: user 1.22 s, sys: 7.95 ms, total: 1.23 s Wall time: 339 ms
```
<details>

Both stop at max_iter in this example, so the result is not affected by the convergence criteria.  If max_iter is increased, scikit-learn converges in 35 iterations annd 3.6 CPU seconds, while statsmodels converges in 17 iterations and 38.8 CPU seconds.

Environment: AWS g4dn instance, AWS Deep Learning AMI Version 29.0
</details>

#### Expected Output

I would expect the run time to be similar when using the same solver and same data.

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

<details>

INSTALLED VERSIONS
------------------
Python: 3.6.10.final.0
OS: Linux 5.3.0-1017-aws #18~18.04.1-Ubuntu SMP Wed Apr 8 15:12:16 UTC 2020 x86_64
byteorder: little
LC_ALL: None
LANG: C.UTF-8

statsmodels
===========

Installed: 0.11.0 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/statsmodels)

Required Dependencies
=====================

cython: 0.29.15 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/Cython)
numpy: 1.18.1 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/numpy)
scipy: 1.4.1 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/scipy)
pandas: 1.0.3 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/pandas)
    dateutil: 2.8.1 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/dateutil)
patsy: 0.5.1 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/patsy)

Optional Dependencies
=====================

matplotlib: 3.1.3 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/matplotlib)
    backend: module://ipykernel.pylab.backend_inline 
cvxopt: 1.2.0 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/cvxopt)
joblib: 0.14.1 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/joblib)

Developer Tools
================

IPython: 7.13.0 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/IPython)
    jinja2: 2.11.1 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/jinja2)
sphinx: 2.4.4 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/sphinx)
    pygments: 2.6.1 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/pygments)
pytest: 5.4.1 (/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/pytest)
virtualenv: Not installed

</details>

**Comment by kshedden** [comment]
Have you tried GLM with binomial family?  It is equivalent to logit, but mostly independent code.

**Comment by josef-pkt** [comment]
try `skip_hessian=True` added as `fit` keyword, that avoids computing the hessian and cov_params.
With 2000 parameters that might take some time

I saw recently a blog comment where statsmodels Logit was faster than sklearn. 
I think it depends on the shape of exog and start_params.

**Comment by aizvorski** [comment]
@kshedden I'm not familiar with GLM, but I just tried this, does this look right?

```python
%time model = sm.GLM(y_train, X_train, family=sm.families.Binomial()).fit(maxiter=10, disp=True)
# CPU times: user 2min 36s, sys: 7.36 s, total: 2min 44s Wall time: 42.2 s
```

Is there any way to do L2 penalty with GLM?

**Comment by aizvorski** [comment]
@josef-pkt 
> try `skip_hessian=True` added as `fit` keyword, that avoids computing the hessian and cov_params.

Okay, I tried it:
```python
%time model = sm.Logit(y_train, X_train).fit(method='lbfgs', pgtol=0.0001, maxiter=10, disp=True, qc_verbose=True, skip_hessian=True)
# CPU times: user 30.9 s, sys: 1.07 s, total: 32 s Wall time: 8.25 s
```

It provides some speedup, but doesn't account for most of the disparity.

**Comment by kshedden** [comment]
For ridge penalty use:

```
sm.GLM(y_train, X_train, family=sm.families.Binomial()).fit_regularized(alpha=###, L1_wt=0)
```

alpha is the ridge penalty weight (alpha=0 is unpenalized)

**Comment by josef-pkt** [comment]
I don't have a very recent sklearn

It looks like statsmodels is spending a lot of time in setting up the model

fit itself is a bit slower per iteration in statsmodels, but converges faster.

```
import sklearn.linear_model
from statsmodels.discrete.discrete_model import Logit
import sklearn.datasets

k = 1000
X_train, y_train = sklearn.datasets.make_classification(n_samples=30000, n_features=k, n_informative=k, n_redundant=0, n_repeated=0)

%time model = Logit(y_train, X_train)
%time model.fit(method='lbfgs', pgtol=0.0001, disp=True, skip_hessian=True)

%time model = sklearn.linear_model.LogisticRegression(solver='lbfgs', penalty='none', verbose=1).fit(X_train, y_train)
Wall time: 4.19 s
Wall time: 956 ms
Wall time: 1.03 s

```

reversing and adding constant to statsmodels X
```
k = 2000
X_train, y_train = sklearn.datasets.make_classification(n_samples=20000, n_features=k, n_informative=k, n_redundant=0, n_repeated=0)
​
%time model = sklearn.linear_model.LogisticRegression(max_iter=100, solver='lbfgs', penalty='none', verbose=1).fit(X_train, y_train)
​
X_train[:, 0] = 1
%time model = Logit(y_train, X_train, has_const=True)
%time model.fit(maxiter=100, method='lbfgs', pgtol=0.0001, disp=True, skip_hessian=True)

Wall time: 1.4 s
Wall time: 4.12 s
Wall time: 1.44 s
```

**Comment by aizvorski** [comment]
@josef-pkt 
> It looks like statsmodels is spending a lot of time in setting up the model

That's a good find!  Yes, that seems to be it, I tested and get the same results.  The setup time scales (roughly) as O(n_samples * n_features**2); if n_features is a bit larger than in my example, this can get really slow

**Comment by josef-pkt** [comment]
I checked with the profile. In the last case nobs=20000, k=2000, most of the time for setting up the model is in np.linalg.svd

eg.
```
%time s = np.linalg.matrix_rank(X_train)
Wall time: 4.41 s
```

**Comment by josef-pkt** [comment]
I'm low in RAM so I don't want to try larger examples, or my chrome and computer freezes.

**Comment by josef-pkt** [comment]
We could add a keyword option to turn off the rank of exog computation, and assume full rank instead.

Currently, that will only affect some degrees of freedom corrections, AFAIR. Those will not make a big difference with a large number of features or large nobs unless the model is a very strange case.

**Comment by alexpghayes** [comment]
A QR decomp would also be faster way to determine the rank, if that's the only place the SVD is being used.

**Comment by josef-pkt** [comment]
Do you know a good way to get the matrix rank using QR?
I know R developers like QR, but we use mostly SVD (by default). I never did any speed comparison.

(I never used a rank revealing, pivoting QR because it didn't exist in python/scipy when we started these parts.)

**Comment by josef-pkt** [comment]
It looks like a rank revealing QR is quadratic in number of columns, but SVD is cubic
e.g. https://core.ac.uk/download/pdf/82831857.pdf p. 78 (based on very selective skimming)

So, for large k as in these examples we would gain.

**Comment by alexpghayes** [comment]
Hmmm didn't realize that the SVD was already truncated. I think QR is generally preferable due to do being more numerically stable than SVD as well. In any case, it's also fairly standard in R for users to be able to turn off rank checking. Really you only get nice handling of low-rank predictor matrices in `stats`, everywhere else it's the wild west.

**Comment by josef-pkt** [comment]
I don't think the SVD is truncated, it still computes all singular values, but it doesn't compute u, v

**Comment by bashtage** [comment]
QR is faster but less numerically accurate. This is why NumPy's `matrix_rank` uses svd (and MATLAB as well).

**Comment by bashtage** [comment]
> Do you know a good way to get the matrix rank using QR?

It is the same as SVD, only you look at the diagonal of the `r` array and compare this to a threshold.  

https://github.com/statsmodels/statsmodels/blob/5618d94eef0f5c09365c89f2260f1231d1a600fe/statsmodels/tsa/deterministic.py#L1256-L1262

**Comment by bashtage** [comment]
```
x = np.random.standard_normal((10000,4000))

%timeit np.linalg.matrix_rank(x)
10.8 s ± 232 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

from scipy.linalg import qr

qr?

 %timeit qr(x, mode="r")
2.36 s ± 24 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
So a good bit faster in the largest case.  Slower in smaller cases though
```
x = np.random.standard_normal((10000,1000))

%timeit qr(x, mode="r")
455 ms ± 12 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit np.linalg.matrix_rank(x)
370 ms ± 10.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

x = np.random.standard_normal((10000,10))

%timeit np.linalg.matrix_rank(x)
242 µs ± 6.41 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit qr(x, mode="r")
1.27 ms ± 22.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by alexpghayes** [comment]
Might also be worthwhile to include some poorly conditioned matrices in your performance tests.

**Comment by bashtage** [comment]
closed by a87938b051a80dc19ff68e954f329791a59f4f6f
